### ISSUE TITLE
macOS Sonoma shows an icon in the menubar with "AltTab is sharing your screen"

### ISSUE THREAD

**Issue by 5cj** [lwouis/alt-tab-macos#2606]
**Describe the bug**

Alert on boot after updating to Sonoma Developer Beta (14.0):
> Alert "AltTab" is using a deprecated API that will be removed in the future. Please contact the developer for an updated version.

**Screenshots / video**

<img width="275" alt="image" src="https://github.com/lwouis/alt-tab-macos/assets/93625793/d0506e4d-b014-4a9b-8c98-f92febf3344d">

**Steps to reproduce the bug**

1. Update to Sonoma Developer Beta
2. Start AltTab
3. Observe alert

**Your environment**

* AltTab version: Version 6.59.0
* macOS version: Sonoma Developer Beta (14.0)

**Comment by lockieluke** [comment]
In the menu bar there’s also an icon indicating that AlrTab is actively monitoring your screen which the icon was not there before the beta 2 update possibly due to newly deprecated APIs

**Comment by 5cj** [comment]
> In the menu bar there’s also an icon indicating that AlrTab is actively monitoring your screen which the icon was not there before the beta 2 update possibly due to newly deprecated APIs

![98FCB2AA-499F-4307-85BB-01C6479D22CC_4_5005_c](https://github.com/lwouis/alt-tab-macos/assets/93625793/706bf9f4-aa48-4324-b91a-166d29f6a8d8)

Screenshot for context, can't find a way to disable or hide it. Unsure if it warrants a separate issue, they may be related.

**Comment by ajnart** [comment]
Same issue here. I've updated to the latest update but have this icon constantly

**Comment by lwouis** [comment]
I already closed these duplicates:

* https://github.com/lwouis/alt-tab-macos/issues/2599
* https://github.com/lwouis/alt-tab-macos/issues/2600
* https://github.com/lwouis/alt-tab-macos/issues/2604

This time, I'll keep this ticket open so people can hopefully see it and stop creating new tickets.

Now regarding the topic itself: a beta is unstable software. Let's wait for Sonoma to go live to discuss if there are things to deal with.

**Comment by lockieluke** [comment]
Interesting take on the issue, a developer beta is for developers to first optimise and fix their apps for the upcoming releases instead of starting to work on it after public release drops

**Comment by lwouis** [comment]
@lockieluke previous macOS beta have introduced similar issues in the past, which they then fixed before the end of the beta. That's what I mean by unstable. Apple develops the OS actively during the beta. For Sonoma, it seems they are working on various UIs to share info about the app to the users. These UI elements look obnoxious and most big apps would display it, so I wouldn't be surprised that it all goes away before the end of the beta.

**Comment by 5cj** [comment]
> Please keep in mind #1179. I want to minimize the work on AltTab's side, so it makes sense to wait for Apple to be done moving things around until I try my best at fixing what they broke.

Fair, but closing issues typically delays progress, especially in a situation like this where it seems Apple is providing a notice. 

For anyone looking for a place to start, I think the deprecated method may be 'activate(ignoringOtherApps:)', though I'm not super confident that's solely responsible for the alert since I have minimal exposure to AppKit development. I found that [Apple's documentation](https://developer.apple.com/documentation/appkit/nsapplication/1428468-activate?changes=latest_minor) has it listed as deprecated in 14.0, and it [appears in the codebase a few times](https://github.com/search?q=repo%3Alwouis%2Falt-tab-macos+ignoringOtherApps&type=code).

**Comment by 0xallie** [comment]
This appears to be an intentionally deprecated feature, not a bug. In fact, it was already deprecated in Ventura, but it could still be used if your app targets a lower version.

> # ATS and ATSUI
> ## Deprecations
> Starting macOS 14 Sonoma, whenever the OS detects the usage of ATS or ATSUI APIs, the user will be presented a dialog stating that the application needs to be updated. Once the dialog is dismissed by the user, the application will exit. (100521621)

https://developer.apple.com/documentation/macos-release-notes/macos-14-release-notes#ATS-and-ATSUI

**Comment by flexlixrup** [comment]
I guess the screen sharing and the deprecated method are 2 different issues. The deprecated API, which throws the error popup has to be fixed before public release of macOS 14. The issue with the screen sharing could be tuned on Apples side before fall.
I suggest opening 2 seperate issues on these topics.

Regarding the deprecated method:
macOS 14 introduces a new 'activate()' method without the flag. The behavior has to be checked as the documentation is minimal but if I would have to guess, ignoringOtherApps is set to false on this new method, so if it changes the behavior of the App, a workaround has to be implemented in time.

**Comment by zacharee** [comment]
It might be possible to use `NSApplicationDelegate.yieldActivation(to: NSRunningApplication)` before `activate()` when switching windows, where the `NSApplicationDelegate` is the app set to be unfocused.

**Comment by zacharee** [comment]
There's a hidden API that supposedly has the same functionality as the deprecated `activate()` API, so I put up a PR implementing it: https://github.com/lwouis/alt-tab-macos/pull/2620.

**Comment by fifafu** [comment]
I have also faced the deprecated API message in my app BetterTouchTool. I'm pretty sure it's not caused by the activateIgnoringOtherApps method. 
Currently the best guess is the CGDisplayStreamCreate API, which is also used in AltTab to check screen recording permissions.

**Comment by koekeishiya** [comment]
I believe you should be using `func CGPreflightScreenCaptureAccess() -> Bool` to check for access,
and `func CGRequestScreenCaptureAccess() -> Bool` to request access if it is not granted already,
to manage screen recording permissions.

https://developer.apple.com/documentation/coregraphics/3656523-cgpreflightscreencaptureaccess
https://developer.apple.com/documentation/coregraphics/3656524-cgrequestscreencaptureaccess

**Comment by fifafu** [comment]
Yes, but this API is not a full replacement. It doesn't update while the app is started and also it is incorrectly marked as available on macOS 10.15 - which will just crash your app if you try to use it on that macOS version.

**Comment by lwouis** [comment]
Absolutely. I tried to use it in the past as it looks on paper like the proper way to do things, but the hacky alternative we use today is more robust and more flexible.

**Comment by caffeinum** [comment]
Not sure if it grants a separate issue, but on MacOS Sonoma, they've also updated UI for a language switcher.

For some reason, AltTab detects it as a separate window, and shows in the app list. It doesn't go away ever, once you switch a language this way once.

**Comment by lwouis** [comment]
Hi @caffeinum,

There was an issue with the new language-switch UI. They fixed it in a later beta release. They will probably continue working on / fixing this feature. Let's review it when Sonoma goes live.

**Comment by Rithari** [comment]
I believe it is safe to say that the issues can be addressed now, seeing as Sonoma goes public Sept. 18th.

**Comment by ThatGuyWhomAsked** [comment]
Correction: Goes live on sept 26, but the release candidate is out so it can be worked on now

**Comment by ThatGuyWhomAsked** [comment]
@lwouis

**Comment by lwouis** [comment]
I tried to create a VM of the Release Candidate 2:

<img width="957" alt="image" src="https://github.com/lwouis/alt-tab-macos/assets/106195/9c35c540-12a7-4948-baa0-0aa70a6d1bdb">

I couldn't get it to start. This means that I'll only be able to experience Sonoma after it's released. Only then, will I be able to observe the issue, and potentially change the app to improve things, if it can be done.

If anyone on Sonoma has an improvement to share, please share a PR and I'll do my best to review it 👍

**Comment by rami3l** [comment]
@lwouis Don’t worry, the official stable release will arrive within this week :)

**Comment by ShlomoCode** [comment]
@lwouis You can easily upgrade by searching in the system settings "Software Update" and then click on the sign next to "Beta updates" and choose "macOS Sonoma Developer Beta" and OK, then you will have the option to upgrade with the click of a button to macOS sonoma (I myself used this method successfully a few days ago )

**Comment by ventsislavnikolov** [comment]
Hi, I'm currently using the RC2 of Sonoma as you can see in the screenshot and I still have the notification for AltTab is recording your screen, just to confirm. And its annoying :)

<img width="460" alt="image" src="https://github.com/lwouis/alt-tab-macos/assets/34811964/e2d47250-b054-40ea-8ca4-5221ba3b5de5">

**Comment by lwouis** [comment]
Anyone knows what this new UI is about?

AltTab needs Screen Recording permissions for its operations. Without it, we can't get screenshots of the windows to show the user. We can't even get the **window names**.

This new UI seems to imply that we are constantly recording the screen, which is not the case. We take screenshots of windows when they change (resized, moved, minimized, etc).

I don't understand what the idea is with this new UI. Maybe it's intended that AltTab is shown there? Maybe Apple's implementation shouldn't trigger here since we are not constantly recording like a video stream? Is there any documentation from Apple somewhere which explains their intent with this new UI system?

**Comment by fifafu** [comment]
I think Apple intents for it to be like this. It goes away after a while when not doing any screen captures for some time. There is no (non-hacky) way to hide that UI.

**Comment by ShlomoCode** [comment]
> This new UI seems to imply that we are constantly recording the screen, which is not the case. We take screenshots of windows when they change (resized, moved, minimized, etc).

Why not just when pressing the alt-tab shortcut?

**Comment by lwouis** [comment]
@ShlomoCode AltTab is refreshing thumbnails when windows move, so that when the user summons it, it can display in the next few frames, with accurate content. We can't capture every change (e.g. a Youtube video playing), so we'll still refresh thumbnails after AltTab shows. But we have something accurate to show right away. Then we update thumbnails one by one from the first to the last. It's typically invisible since often all thumbnails are already up to date.

So in short, we do that to provide the best UX. AltTab aims at being as responsive as it is on Windows, which is a high target to aim at.

**Comment by Rithari** [comment]
Sonoma is available starting today.

https://www.apple.com/newsroom/2023/09/macos-sonoma-is-available-today/

**Comment by mikluko** [comment]
Screen recording icon in control center is tremendously annoying. I know there is probably nothing that can be done on the developer side but I can't afford attention drain like that. 

😭 I hate to do that but I have to abandon AltTab.

**Comment by HamzaM** [comment]
> Screen recording icon in control center is tremendously annoying

Got the same issue with the stable release of Sonoma, quite annoying...

**Comment by emirrou** [comment]
> Screen recording icon in control center is tremendously annoying. I know there is probably nothing that can be done on the developer side but I can't afford attention drain like that.
> 
> 😭 I hate to do that but I have to abandon AltTab.

You can also check the option "Hide window thumbnails" which stops showing the screen recording icon. Makes AltTab still work just fine :)

**Comment by momos11** [comment]
It seems like this has something to do with the used API. There should be a new API which doesn't have this problem.

**Comment by rami3l** [comment]
alt-tab is not alone in this situation though. With certain options turned on, Bartender 5 is doing the same.

**Comment by Rithari** [comment]
> alt-tab is not alone in this situation though. With certain options turned on, Bartender 5 is doing the same.

Might be worth talking to their development team to see what they've come up with considering Bartender 4 faced the same issues by default.

**Comment by ldclassic** [comment]
Maybe also checkout the window switcher in BetterTouchTool (predefined action "Show Window Switcher For All Open Apps"). It seems to be able to show thumbnails of windows without triggering the screen recording icon (also in general seems to have similar functionality to AltTab).

**Comment by lwouis** [comment]
Everyone keeps calling the UI an "issue". As far as i understand, it's exactly what Apple wants: make people extra aware of apps capturing their screen or windows.

i think it's a feature not a bug, and people who don't like to be notified so boldly may want to turn to Apple. Apple designed this warning UI. It seems to me it's behaving exactly as they intended. AltTab can't control what Apple does here

**Comment by Rithari** [comment]
I believe it to be an issue effectively because this change will make people not want to use great apps such as AltTab or Bartender (although the latter mitigated the issue by providing a viable alternative) and disabling the thumbnails might not be everyone's cup of tea. 

That being said, I do agree that it's just Apple being Apple thing and hardly your fault here, as far as the change goes.

**Comment by starcraft66** [comment]
@lwouis While I mostly agree with your sentiment and the spirit of the feature, by not allowing users to disable it for certain apps or altogether, Apple is just subjecting its users to alert fatigue.
If this indicator is supposed to alert me of privacy problems, but is firing up with a 99.99% false-positive rate, I will just start ignoring it and not notice any privacy problems it should actually be alerting me of.

**Comment by Duc-NSH** [comment]
> @lwouis While I mostly agree with your sentiment and the spirit of the feature, by not allowing users to disable it for certain apps or altogether, Apple is just subjecting its users to alert fatigue. If this indicator is supposed to alert me of privacy problems, but is firing up with a 99.99% false-positive rate, I will just start ignoring it and not notice any privacy problems it should actually be alerting me of.

Totally agree with you, hope that Apple will bring an option to disable the warning for specific applications. Currently, I have to disable the thumbnail temporarily :(

**Comment by ourcore** [comment]
To me, it's not a big enough dealbreaker to disable thumbnails since it only appears in the menu bar temporarily, I just wish I could at least hide it with other icons, but like others have said, it appears to be doing what Apple intended and hiding it would probably require a hack-y solution. Apple would never let users disable it

**Comment by ShlomoCode** [comment]
You can drag it using command beyond the visible amount of icons (if you have many apps with an icon in the menubar)

**Comment by ourcore** [comment]
The point of the discussion is that there might not be an issue to resolve...

**Comment by mikluko** [comment]
> You can also check the option "Hide window thumbnails" which stops showing the screen recording icon. Makes AltTab still work just fine :)

For me like 90% of app value was exactly window previews. Without them built-in switching works better at my taste.

**Comment by mikluko** [comment]
> Everyone keeps calling the UI an "issue". As far as i understand, it's exactly what Apple wants: make people extra aware of apps capturing their screen or windows.
> 
> i think it's a feature not a bug, and people who don't like to be notified so boldly may want to turn to Apple. Apple designed this warning UI. It seems to me it's behaving exactly as they intended. AltTab can't control what Apple does here

From purely engineering standpoint you are right since everything works as intended and expected. But from usability standpoint it is a degradation of experience and constitutes a proper issue.

**Comment by lwouis** [comment]
@mikluko please share this feedback with Apple. Apple pushed this change to all their customers.

**Comment by starcraft66** [comment]
I filed FB13209646 with more or less the content of my previous reply.

**Comment by iammattlewis** [comment]
> Everyone keeps calling the UI an "issue". As far as i understand, it's exactly what Apple wants: make people extra aware of apps capturing their screen or windows.
> 
> i think it's a feature not a bug, and people who don't like to be notified so boldly may want to turn to Apple. Apple designed this warning UI. It seems to me it's behaving exactly as they intended. AltTab can't control what Apple does here

Yes, this is the "issue". This is not an AltTab bug, but rather a Sonoma feature. I have contacted Apple, but they haven't provided any update/resolution as of yet. It's not a bad thing to have this feature, but not allowing the user to turn it off/turn it off for certain apps is not very user friendly.

**Comment by iammattlewis** [comment]
Conversation with Apple Support. Also had a phone conversation with support team member. Gist of the call was "it's a privacy feature, but it shouldn't be showing up that much." I explained I was using a third party window switcher, and he said there is currently not a way to stop it, and there are no plans to implement the ability to turn off the menu bar notification. 

Unfortunately, this will discontinue my use of AltTab for the time being, and that is entirely the fault of Apple. I wish they would let me use my computer how I want instead of implementing features that you have zero control over, but alas...that is the Apple way, it seems.

<img width="595" alt="image" src="https://github.com/lwouis/alt-tab-macos/assets/118700125/28398ae9-3839-4f48-a0ba-b1065b19d45e">

**Comment by arthurflor23** [comment]
I would suggest an improvement on AltTab without thumbnails by default, because macos alt+tab appearance is good, but it doesn't allow as many customizations as AltTab. So it could be a possibility to take advantage of.

**Comment by ravibrock** [comment]
> You will also notice a new menu bar item in macOS Sonoma, informing you that Bartender is capturing your screen whenever it gets an image of your menu bar items, again this is new in macOS Sonoma, and currently unavoidable, because of the same API issue. I am working on Bartender 5 right now that will have the ability to avoid this happening.

(From [Bartender's website](https://www.macbartender.com/Bartender4/support/))

It seems there could be a workaround? I'm sorry that I don't know the details, but perhaps it could work for AltTab as well.

**Comment by fungiboletus** [comment]
As I was sharing my screen in a meeting, I noticed that the new screen capturing notification wasn't bothering me near as much as when it appears and disappear every time I use AltTab.

So behold the following terrible hack to make the purple icon always visible:

```lua
function triggerScreenCaptureNotification()
  local allWindows = hs.window.allWindows()

  local smallestWindow = nil
  local smallestSize = 100000000

  for _, window in pairs(allWindows) do
    local currentSizeObj = window:size()
    local currentSize = currentSizeObj.w * currentSizeObj.h
    if currentSize < smallestSize then
      smallestSize = currentSize
      smallestWindow = window
    end
  end

  local snapshot = smallestWindow:snapshot()
end

triggerScreenCaptureNotification()
timerScreenCapture = hs.timer.doEvery(9, triggerScreenCaptureNotification)
```

Put that in your [HammerSpoon](http://hammerspoon.org) configuration and allow it to screen capture your screen.

I am not responsible if someone captures your screen and you don't notice. Use at your own risk. Code is released under the [WTFPL](http://www.wtfpl.net).

**Comment by SarcevicAntonio** [comment]
Can we make the "Hide window thumbnails" option show the App icons instead, nice and big, like for apps with no open windows?

**Comment by z4ffe** [comment]
@SarcevicAntonio Good idea!

**Comment by Lakshmann** [comment]
That sounds fair: leave the thumbnail previews behind but still have the window-based switching with a look closer to the native cmd-tab (larger icons, as stated above).

**Comment by aubreypwd** [comment]
> AltTab is refreshing thumbnails when windows move, so that when the user summons it, it can display in the next few frames, with accurate content.

I can confirm that with this off:

<img width="213" alt="image" src="https://github.com/lwouis/alt-tab-macos/assets/1753298/56e15cdf-d5ad-4a86-938a-5b1b4b6d608c">

I don't see the purple button anymore.

**Comment by iammattlewis** [comment]
Using it without the window thumbnails is fine, but defeats a lot of the purpose (at least for me).

**Comment by conker-rsc** [comment]
> Can we make the "Hide window thumbnails" option show the App icons instead, nice and big, like for apps with no open windows?

+1 to that idea -- with this option enabled, one could also set "window app icon size" to 0px to get even more room for the window title above the icon, which would be nice as well.

**Comment by MaxCollinYuan** [comment]
Bartender 5 has made it possible not to display the purple button "XXX is sharing your screen" on the menu bar. Why can't AltTab do it?

**Comment by lwouis** [comment]
@MaxCollinYuan the 2 apps do different things. I guess they found a way to not screenshot the menubar. Maybe Sonoma introduced new APIs to inspect the menubar?

AltTab has to screenshot the windows to show thumbnails to the user. We also need Screen Recording permission to get the window titles. See #1082 to learn more.

The way you ask your question is rude. Some other people in this thread are rude as well. AltTab is me spending probably 300h at this point, on my weekends and evenings, to create an app and share it with other people. I don't take donations and the only thing I get from this project is the satisfaction of helping other people.

Imagine going to a dog shelter, and complaining to the guy who runs it about some inconvenience you experienced in the shelter. How entitled are you? If you don't like something about this free app, the decent behavior is to politely ask if it can be improved. There is also the option to not use the free app and be silent.

It's not often i have to deal with unpleasant people, thankfully. This Sonoma issue has brought a crowd most unpleasant for some reason.

If you want a change, advocate to Apple. Or share a pull-request with your own work to improve things. Or ask people here politely if they could help you with a fix. Or uninstall the app and have the dignity to not complain about something you got for free in the first place.

Thank you

**Comment by emirrou** [comment]
> Bartender 5 has made it possible not to display the purple button "XXX is sharing your screen" on the menu bar. Why can't AltTab do it?

As of right now bartender does trigger the purple icon, but only when you use the "Bartender Bar", which shows your menu bar items below the actual menu bar. iirc in bartender5 it is turned off by default and in the settings it also warns you about this behaviour.

**Comment by ayroblu** [comment]
For what it's worth, I submitted feedback to Apple via their feedback assistant, I'm not really sure if it's the right place or if they even look at it, but just sharing here in case anyone wants to use it as a template for feedback too, as I'm sure more reports would lead to more traction on their side.

> ### Basic Information
> Please provide a descriptive title for your feedback:

Disable \<AppName> is sharing your screen for certain apps due to security risk

> Which area are you seeing an issue with?

Menu Bar

> What type of issue are you reporting?

Incorrect/Unexpected Behavior

> ### Details

> What does the Menu Bar issue you are seeing involve?

Something else not on this list

> What is the active application?

AltTab

> Which display is the menu bar on?

My primary display

> ### Description
> Please describe the issue and what steps we can take to reproduce it:

When using known accessibility tools like AltTab, macOS shows a "AltTab is sharing your screen". (details here: https://alt-tab-macos.netlify.app/) This is a nice new feature generally, but it's only half implemented. It's currently implemented in a somewhat useless way cause there's no way to exclude known apps. This means that users see it continuously, and will therefore ignore it. This poses a security risk if users learn to ignore it and a malicious app starts doing this in the background.

**Comment by sapoepsilon** [comment]
> > You will also notice a new menu bar item in macOS Sonoma, informing you that Bartender is capturing your screen whenever it gets an image of your menu bar items, again this is new in macOS Sonoma, and currently unavoidable, because of the same API issue. I am working on Bartender 5 right now that will have the ability to avoid this happening.
> 
> (From [Bartender's website](https://www.macbartender.com/Bartender4/support/))
> 
> It seems there could be a workaround? I'm sorry that I don't know the details, but it might work for AltTab as well.

For what it's worth, Bartender is a $16 app, while AltTab is a free app. It's probable that a workaround could be implemented on AltTab as well, but that might take a considerable amount of time. Developers contribute here whenever they have free time away from their primary jobs or school. They contribute because they enjoy development, and pursue it as a hobby.

For everyone who is condescending towards the main developer of this project, lwouis, recall what Korolev once stated:
 > "If you do not agree — criticize; if you criticize — offer a solution/alternative; if you offer a solution — work on implementing it."

**Comment by anglee0323** [comment]
> @MaxCollinYuan the 2 apps do different things. I guess they found a way to not screenshot the menubar. Maybe Sonoma introduced new APIs to inspect the menubar?这两个应用程序做不同的事情。我猜他们找到了一种不截图菜单栏的方法。也许索诺玛引入了新的 API 来检查菜单栏？
> 
> AltTab has to screenshot the windows to show thumbnails to the user. We also need Screen Recording permission to get the window titles. See #1082 to learn more.AltTab 必须对窗口进行屏幕截图才能向用户显示缩略图。我们还需要屏幕录制权限才能获取窗口标题。请参阅#1082 了解更多信息。
> 
> The way you ask your question is rude. Some other people in this thread are rude as well. AltTab is me spending probably 300h at this point, on my weekends and evenings, to create an app and share it with other people. I don't take donations and the only thing I get from this project is the satisfaction of helping other people.你问问题的方式很粗鲁。这个线程中的其他一些人也很粗鲁。 AltTab 是我在周末和晚上花费了大约 300 个小时来创建一个应用程序并与其他人分享。我不接受捐款，我从这个项目中得到的唯一的东西就是帮助别人的满足感。
> 
> Imagine going to a dog shelter, and complaining to the guy who runs it about some inconvenience you experienced in the shelter. How entitled are you? If you don't like something about this free app, the decent behavior is to politely ask if it can be improved. There is also the option to not use the free app and be silent.想象一下，你去了一个狗收容所，并向收容所的经营者抱怨你在收容所遇到的一些不便。你有什么权利？如果您不喜欢这个免费应用程序的某些内容，礼貌的行为是礼貌地询问是否可以改进。还可以选择不使用免费应用程序并保持沉默。
> 
> It's not often i have to deal with unpleasant people, thankfully. This Sonoma issue has brought a crowd most unpleasant for some reason.谢天谢地，我并不经常与不愉快的人打交道。出于某种原因，索诺玛的这个问题给人群带来了最不愉快的感觉。
> 
> If you want a change, advocate to Apple. Or share a pull-request with your own work to improve things. Or ask people here politely if they could help you with a fix. Or uninstall the app and have the dignity to not complain about something you got for free in the first place.如果你想改变，请向苹果公司求助。或者与您自己的工作共享拉取请求以改进内容。或者礼貌地询问这里的人是否可以帮助您解决问题。或者卸载该应用程序，并有尊严地不抱怨你一开始就免费获得的东西。
> 
> Thank you 谢谢

I understand how you feel. Some people do tend to talk too much. I hope you don't take it personally. Lately, I have been looking into Apple's new API, and I have formed some opinions on how to solve this problem. I hope those who have a desire to contribute can work together to improve this poor user experience. Wishing you a happy life!

**Comment by Philippe-cheype** [comment]
> Can we make the "Hide window thumbnails" option show the App icons instead, nice and big, like for apps with no open windows?

I've always been used to the way Windows showed the preview of the app, I found that to be a very useful feature, but It's very annoying to have that purple icon at the top, and as @ayroblu puts it, is it a security risk.

I could get used to the native ⌘+Tab look, though, I would still use AltTab, as the native one is poorly thought out, and gets messy very quickly. Doing this on AltTab is actually almost possible, just by playing with the settings. The only piece missing (I guess) is that 0px titles should take 0 width (or there should be an option to disable titles):

<img width="562" alt="Screenshot 2023-10-02 at 4 00 17 PM" src="https://github.com/lwouis/alt-tab-macos/assets/72014356/0031c7b1-3b13-43a4-b3a4-01e620c5d79d">

My settings:

<img width="592" alt="Screenshot 2023-10-02 at 4 00 13 PM" src="https://github.com/lwouis/alt-tab-macos/assets/72014356/5a6ceee5-03d1-49f6-97b1-c260a0765510">

**Comment by lwouis** [comment]
@Philippe-cheype You can get closer to the native switcher look by tweaking the sliders more. `Window min width in row` in your case especially, I think. Some people have achieved essentially the same look at the native switcher in the past.

There is an ongoing effort here to make the switch between "Window alt-tab style" and "native app switcher style" very easy. There is more to it actually. Lots to read here: #351 and the open PR https://github.com/lwouis/alt-tab-macos/pull/2697.

**Comment by Philippe-cheype** [comment]
@lwouis OH! that's amazing. Great work from all the contributors, AltTab really does it all.

(For anyone interested: the settings I showed above + `Window min width in row`:  `1%` and play with the `Window app icon size` to your liking)

**Comment by Qhilm** [comment]
> It's probable that a workaround could be implemented on AltTab as well, but that might take a considerable amount of time.

I don't think so. From apple's point of view, it's a security feature. If there was a workaround allowing not to display the dreaded icon in the menu bar while actually capturing element of the GUI of macOS (windows, menu bar, etc.), that would be considered a vulnerability.

For this reason I think Apple developed this "feature" in a way that it is not possible to work your way around it – unless you actually hack macOS sonoma, but I think that would put altTab on a whole different plane, a much riskier one for the developer actually (and also significantly more time consuming).

Hence I think the recommendation to voice your discontent with Apple is the correct feedback here, as it's the only realistic way forward: Apple must implement in macOS a way to whitelist trusted (and code signed) apps.

I'll be honest though and say my hopes are very low on this front...

**Comment by Damecek** [comment]
If the Screen Capture feature would not be mandatory we will loose the thumbnails + window titles. The thumbnails is not a huge deal as you can turn on the preview of selected window. Personally I would give up on the window titles if it would mean we can get rid of the purple icon in bar.

**Comment by lwouis** [comment]
> The thumbnails is not a huge deal as you can turn on the preview of selected window.

The preview is also a screenshot of the window. It would trigger the Sonoma warning like thumbnails do.

**Comment by Damecek** [comment]
> The preview is also a screenshot of the window. It would trigger the Sonoma warning like thumbnails do.
:/ i thought that it just bring the window to foreground like focus. that is unfortunate.

**Comment by Qhilm** [comment]
From the [bartender website](https://www.macbartender.com/Bartender5/ScreenCaptureMenuBarItem/):

>Whenever Bartender gets an image of a menu bar item macOS will show the purple screen capture menu bar item

>This menu bar item cannot be hidden, or disabled by Bartender and is put there by macOS

Also, I own bartender 5 and I can confirm the above is accurate for bartender 5 (latest version).

Bartender "simply" added a switch to disable the feature triggering the security warning, they didn't really worked around it in my view. Bartender is lucky this feature is not core to its functionality.

Unfortunately, a window switcher is a different type of app and the snapshot/preview of the windows is a core feature, removing it makes the app unusable (at least for me): I would be blindly switching around between 10 windows all represented by the same Firefox icon – if I have 10 Firefox windows open, for example. This is not usable.

My view is, unless there's a change in macOS, it is not possible for AltTab to prevent this icon from showing while maintaining a usable app. In the mean time, there are only two options: stop using a window switcher on macOS (not just AltTab but any window switcher) or get used to the purple icon in the menu bar...

**Comment by Neoconair** [comment]
Not a solution but using just the app icons + window titles has acutally proved a better way to use the app for me.

Turns out I can process the titles faster than the window previews!

**Comment by Qhilm** [comment]
Nice, I just realised, I can make AltTab behave like Witch (which I gave up on because it was too slow – haven't used it in a long while though).

I will give it a try but I'm worried about Firefox (to use my prior example): window title is the active tab. I use a window for projects, and they tend to have tabs to same websites open. Let's try it.

Thanks!

**Comment by jhonalves** [comment]
It has been a week since i updated to macOS Sonoma and now i'm used to the icon in the menu bar. It's still a little annoying, but most of the time i don't even notice it.
My experience reinforce what was said here that Apple got it wrong when they did not provide an option to create a whitelist with trusted apps and now the user will always ignore the warning, even if it's not from a trusted app.

**Comment by andres-martinez-torres** [comment]
It would be awesome to keep the coloured circles when we disable the window thumbnails. I personally use the feature a lot.

**Comment by siikdUde** [comment]
Seeing as how this probably won't be fixed anytime soon, is there at least an option to remove alt tab from displaying on the menu bar? A lot of other apps have the option to remove it and I certainly don't need to see it as it takes up space for other more useful menu bar items like google drive, dropbox, etc. Can this option be added please?

**Comment by sapoepsilon** [comment]
> Seeing as how this probably won't be fixed anytime soon, is there at least an option to remove alt tab from displaying on the menu bar? A lot of other apps have the option to remove it and I certainly don't need to see it as it takes up space for other more useful menu bar items like google drive, dropbox, etc. Can this option be added please?

It is already there. Check in the settings window

**Comment by siikdUde** [comment]
> > Seeing as how this probably won't be fixed anytime soon, is there at least an option to remove alt tab from displaying on the menu bar? A lot of other apps have the option to remove it and I certainly don't need to see it as it takes up space for other more useful menu bar items like google drive, dropbox, etc. Can this option be added please?
> 
> It is already there. Check in the settings window

I have many times for the past few months on all categories. I must be missing something then. Would you be able to give a screenshot of where it is exactly in the settings? I would really appreciate it, thank you

**Comment by sapoepsilon** [comment]
> > > Seeing as how this probably won't be fixed anytime soon, is there at least an option to remove alt tab from displaying on the menu bar? A lot of other apps have the option to remove it and I certainly don't need to see it as it takes up space for other more useful menu bar items like google drive, dropbox, etc. Can this option be added please?
> > 
> > 
> > It is already there. Check in the settings window
> 
> I have many times for the past few months on all categories. I must be missing something then. Would you be able to give a screenshot of where it is exactly in the settings? I would really appreciate it, thank you

![image](https://github.com/lwouis/alt-tab-macos/assets/47342870/ca03e81f-fe6b-46aa-98fb-e32f87a891bf)

**Comment by siikdUde** [comment]
Wow, I completely just overlooked that. Thanks so much!

**Comment by xconverge** [comment]
I still see the icon every once in awhile with window previews disabled (where it was shown everytime I alt-tabbed)

Does anyone know what feature may be causing this and if it is able to be disabled as well?

**Comment by lwouis** [comment]
@xconverge it may be the periodic check that the screen recording permission is given. The idea is to make a 0x0 screenshot to confirm that the permission is granted. There is no better API/way to check that.

It could also be when we read window titles.

I'm still not on Sonoma, so i haven't dug into the details of this new UI, and which exact API calls trigger it.

**Comment by HackettLai** [comment]
<img width="588" alt="Screenshot 2023-10-06 at 09 39 27" src="https://github.com/lwouis/alt-tab-macos/assets/56993446/1f3e51f4-5b34-4d02-ae59-b840ab96931b">
![vlcsnap-2023-10-06-09h40m12s286](https://github.com/lwouis/alt-tab-macos/assets/56993446/f1eeb230-d9fd-47ed-89b9-22dfe2a0f460)

I found this setting work great on me, just sharing with all you guys

**Comment by Qhilm** [comment]
> coloured circles

What are these coloured circles? What do they do?

**Comment by andres-martinez-torres** [comment]
> > coloured circles
> 
> What are these coloured circles? What do they do?

Basically the red, yellow, green buttons for each window (close, minimise, full-screen). I often used those to close windows in alt-tab without having to open them.

**Comment by Duc-NSH** [comment]
> <img alt="Screenshot 2023-10-06 at 09 39 27" width="588" src="https://user-images.githubusercontent.com/56993446/273081914-1f3e51f4-5b34-4d02-ae59-b840ab96931b.png">
> ![vlcsnap-2023-10-06-09h40m12s286](https://user-images.githubusercontent.com/56993446/273082087-94a34d85-41e4-418d-8e2b-ee0e8f1986e5.png)
> 
> I found this setting works great for me, just sharing it with all of you guys

This setting is great for me too, now I can use alt-tab without thumbnail comfortably

**Comment by lwouis** [comment]
I've updated to Sonoma today, and was able to investigate this.

It seems to only show when we make the API call to take a screenshot. It doesn't seem to show the purple indicator when we:

* Get a window title
* Get an app icon
* Get window metadata (e.g. minimized, hidden, dimensions, etc)

The only thing we could do, it seems to me, is offer an AltTab experience without thumbnails, for people who can't stand the purple icon. We already do that by checking out the checkbox in `Preferences > Appearance > Hide window thumbnails`.

There are 2 tickets which may improve and expand upon that thumbnail-less experience:

* #351 
* #1082 

Closing this ticket as there is no direct action we can take, and incremental improvements are tracked in the tickets above.

**Comment by wookayin** [comment]
>  It doesn't seem to show the purple indicator when we:
> 
> Get a window title
> Get an app icon
> Get window metadata (e.g. minimized, hidden, dimensions, etc)

@lwouis Thanks for the support, but I think there still are some cases where AltTab is taking screenshots even when it's not needed for thumbnails.

I use AltTab for `cmd-` (shows screenshot/thumbnail) and `option-tab` only: for me, `cmd-tab` is the macOS builtin cmd-tab. So when I do `cmd-tab`, AltTab is neither supposed to work, nor should try to capture window metadata or app icon, etc. But I can *consistently* see that whenever I open the *macOS built-in* window switcher with `cmd-tab`, the purple icon in the menubar apperas saying that "AltTab is capturing your screen". So I suspect there still remains a case where AltTab is trying to take screenshot even when it's not really needed.

If you think this is a separate bug that you'd like to track outside this thread, I'd be happy to open a new issue.

**Comment by lwouis** [comment]
Hi @wookayin,

AltTab monitors changes in windows, and when important changes happen, it takes a new screenshot of the window. This happens in the background, in addition to when AltTab UI is showing. For example, if you create a new window, AltTab will screenshot it, in the background. That way, when you next open AltTab, it can show extremely quickly, since it's ready with the screenshots of every window. No delay to get the screenshots ready.

Try to check the `Preferences > Appearance > Hide window thumbnails`. At this point, I don't see you'll ever see the purple icon again. If you do, then yes, that would indicate something else I'm not thinking of is going on.

**Comment by xconverge** [comment]
I definitely saw the same behavior with the disable thumbnail button checked as I mentioned up above so I think there is something else going on

**Comment by lwouis** [comment]
@xconverge could you please open a ticket with steps to reproduce and a video showcasing the issue?

Thank you

**Comment by ElhemEnohpi** [comment]
@Qhilm -

> I will give it a try but I'm worried about Firefox (to use my prior example): window title is the active tab. I use a window for projects, and they tend to have tabs to same websites open. Let's try it.

I also work that way in Firefox. I use the "Window Titler" addon. It lets you give a custom title to each window, i.e. the name of the project, and it's prepended to whatever the active tab title is. I turned off thumbnails because of the purple icon issue, and re-arranged the appearance to a vertical list of the titles with a single column. I kind of miss the thumbnails, but I can still use the important functions of Alt-Tab about as well as before. There's even the advantage that I can see more of the titles now.

**Comment by scottkuhl** [comment]
> It's stupid, there should at least be an option to exclude apps from it - especially as it doesn't improve security at all. However this is feedback that needs to go to Apple. I filed it as FB13205878

This is the best answer given.  Apple has introduced a security feature, so users see when something is trying to record their screen.  No idea if that will ever make a single difference, but I believe this is the reason.  There absolutely should be a way for users to mark an app as safe.

The option to turn off the thumbnails is a nice workaround but disables a great feature of AltTab.  There is nothing more this project can do, so pressure should be put on Apple.

@fifafu is there any link to that feedback to Apple to track status or pile on votes?

**Comment by burdo3417** [comment]
Updated to Sonoma 14.1 today and the Warning icon in the menu bar seems to be gone. All good even whith thumbnails on.

**Comment by conker-rsc** [comment]
> Updated to Sonoma 14.1 today and the Warning icon in the menu bar seems to be gone. All good even whith thumbnails on.

I am still seeing it occasionally - just far less often than 14.0

**Comment by Duc-NSH** [comment]
> > Updated to Sonoma 14.1 today and the Warning icon in the menu bar seems to be gone. All good even whith thumbnails on.
> 
> I am still seeing it occasionally - just far less often than 14.0

same here, it does not always show as in 14.0 but it still shows from time to time

**Comment by DownrightNifty** [comment]
Just recently upgraded to Sonoma. The icon is distracting and it's frustrating that Apple provides no option to disable it. Regardless, I appreciate the effort that went into compensating for Apple's bad decisions.

I was recently thinking about this, though: would it be less distracting if the icon was permanently there? Maybe there could be an option to make the app continually make bogus API requests so the icon never goes away?

**Comment by burdo3417** [comment]
> > > Updated to Sonoma 14.1 today and the Warning icon in the menu bar seems to be gone. All good even whith thumbnails on.
> > 
> > 
> > I am still seeing it occasionally - just far less often than 14.0
> 
> same here, it does not always show as in 14.0 but it still shows from time to time

I haven't seen it again. Not once. Weird.

**Comment by ThatGuyWhomAsked** [comment]
Anyone know why I haven't seen it? (Not complaining)

**Comment by ThatGuyWhomAsked** [comment]
> > > > Updated to Sonoma 14.1 today and the Warning icon in the menu bar seems to be gone. All good even whith thumbnails on.
> > > 
> > > 
> > > I am still seeing it occasionally - just far less often than 14.0
> > 
> > 
> > same here, it does not always show as in 14.0 but it still shows from time to time
> 
> I haven't seen it again. Not once. Weird.

did they change the api? Im curious as to why apps like bartender still do it

**Comment by Qhilm** [comment]
> I was recently thinking about this, though: would it be less distracting if the icon was permanently there, rather than popping up intermittently? Maybe there could be an option to make the app continually make bogus API requests so the icon never goes away?

I think it defeats the whole purpose of this macOS security feature. If an a malicious app starts recording your screen, you will never notice, because the icon is anyway displayed around the clock on your device.

The purpose of this (admittedly) annoying icon is to alert the user that some app is recording a portion or the whole screen.

The right way to fix this is for apple to implement whitelisting. But unless the product of a major player (adobe, microsoft, google, etc.) is impacted, I don't see Apple implementing it.

**Comment by fifafu** [comment]
> > I was recently thinking about this, though: would it be less distracting if the icon was permanently there, rather than popping up intermittently? Maybe there could be an option to make the app continually make bogus API requests so the icon never goes away?
> 
> I think it defeats the whole purpose of this macOS security feature. If an a malicious app starts recording your screen, you will never notice, because the icon is anyway displayed around the clock on your device.
> 
> The purpose of this (admittedly) annoying icon is to alert the user that some app is recording a portion or the whole screen.
> 
> The right way to fix this is for apple to implement whitelisting. But unless the product of a major player (adobe, microsoft, google, etc.) is impacted, I don't see Apple implementing it.

As long as malicious apps can just overlay that icon with something else, this is not a security feature. Period.

**Comment by Qhilm** [comment]
I disagree. Security is not black and white, it's layers of swiss cheese. The difficulty level to overlay something on this icon in a reliable manner will be significantly higher compared to "simply" capturing the screen.

It's akin to saying "because someone can walk into my DC, it's not worth putting up firewalls".

**Comment by Qhilm** [comment]
If it's really that easy, then I would tend to agree with you, it's not very useful and even rather detrimental. My understanding was that it was rather hard to reliably hide a menu bar item, as it shifts around depending on what is displayed in the menu bar...

**Comment by user334** [comment]
So how has it been so far on 14.1/14.1.1/14.2? Does macOS still show this purple icon while using AltTab or not?

**Comment by mgc8** [comment]
> So how has it been so far on 14.1/14.1.1/14.2? Does macOS still show this purple icon while using AltTab or not?

Just upgraded to 14.1.1 and it does happen, so far only noticed it with the Alt-Tab app. It comes up with a delay, usually a couple of seconds after pressing the keyboard shortcut, then disappears after another delay. Sometimes it goes away completely for a while, then comes back later out of the blue. It's _very_ inconsistent, which makes it _extremely_ annoying -- the purple thing screams in your peripheral vision while trying to do something different, interrupting thoughts and workflows.

I wish there was a way to just say "thanks, go away" or disable this misfeature completely, but alas big daddy Apple knows better :-/

**Comment by DownrightNifty** [comment]
> I think it defeats the whole purpose of this macOS security feature. If an a malicious app starts recording your screen, you will never notice, because the icon is anyway displayed around the clock on your device.

Yes, I understand that. The proposed option would be disabled by default. If you are concerned about this, then you shouldn't enable the option. In my case, AltTab is the only app I've granted permission to access my screen, so there is no ambiguity when the icon displays. With the option enabled, you should view the "screen recording" permission the same way you did pre-Sonoma: it allows an app to record your screen in the background at any time. If you want, you can temporarily provide access to an app (e.g. for a video call) and then revoke it when you're done.

> The right way to fix this is for apple to implement whitelisting.

Yes, but we have little to no influence over what Apple does.

AltTab is a special case among apps that need screen recording permission; most only make use of that access intermittently, but AltTab accesses it constantly. Apple has failed to accommodate for this type of app when designing their security controls. Until they fix this, hacky workarounds are the only solution unfortunately.

Users will always have the ability to use an external tool for this; making the indicator constantly show is not an "exploit" in and of itself. You can download and use the example provided by fungiboletus right now if you want. Implementing it within AltTab just makes it easier and more memory-efficient.

**Comment by DownrightNifty** [comment]
I'm now on 14.1.1. My experience is similar to other reports here: the icon now mysteriously shows from time to time. Shows less often, but inconsistent. Still quite annoying.

But I noticed something interesting. Hammerspoon seems to be able to capture windows without tripping the indicator. I used the [config provided by fungiboletus](https://github.com/lwouis/alt-tab-macos/issues/2606#issuecomment-1739342742), which was designed specifically to make the indicator show, but it doesn't seem to ever appear on 14.1.1. I've even tried restarting the app and computer multiple times. They might be using an API that doesn't trip it somehow. This should be investigated.

**Comment by onurcagan** [comment]
Just checking in to see if there's any work on this as this is quite annoying. I can also look into this if @DownrightNifty's solution seems to be able to achieve what everyone wants.

**Comment by lwouis** [comment]
`CGWindowListCreateImage` is what we used to use before switching to the private API . Please see https://github.com/lwouis/alt-tab-macos/blob/9e03d5f6659f0bf2327712e1e70851267acf01c1/src/api-wrappers/CGWindowID.swift#L29

I would also be surprised it would not show the indicator. Or at least differ from other APIs like the one we use

**Comment by DownrightNifty** [comment]
@lwouis I compiled AltTab with those APIs swapped out and the indicator still shows up (inconsistently).

It's really weird that Hammerspoon doesn't trip the indicator at all though. I haven't gotten the time to do an in-depth investigation, but I figure it might have something to do with the frequency of the API calls?

Based on the tip from @fifafu that the code responsible for this is located in Control Center, I'd like to pop `/System/Library/CoreServices/ControlCenter.app/Contents/MacOS/ControlCenter` into Ghidra, if I get the time before school starts back up for me...

**Comment by DownrightNifty** [comment]
If you [disable SIP](https://developer.apple.com/documentation/security/disabling_and_enabling_system_integrity_protection), you could actually make patches to the ControlCenter binary to disable the indicator for AltTab, with root access. This of course opens up security vulnerabilities but if the indicator annoys you that bad, maybe it's worth it, I don't know. I probably wouldn't do it myself. (Though I have temporarily disabled SIP in the past to do dynamic analysis on macOS system binaries 😅.)

**Comment by DownrightNifty** [comment]
```
$ strings /System/Library/CoreServices/ControlCenter.app/Contents/MacOS/ControlCenter | grep "capturing your screen"
 is capturing your screen.
```

Just based on this basic smoke test, it seems very likely that this is indeed the binary responsible for the indicator :)

**Comment by SarcevicAntonio** [comment]
Not sure what happened, but I'm running macOS 14.2.1 and right now the indicator doesn't seem to appear anymore when using AltTab.

**Comment by Duc-NSH** [comment]
> Not sure what happened, but I'm running macOS 14.2.1 and right now the indicator doesn't seem to appear anymore when using AltTab.

Hmm, it is weird, I am currently on 14.2.1 too but the indicator still appears occasionally

**Comment by lwouis** [comment]
As someone mentioned in that thread, i believe that Apple has patched their purple indicator over time.

It used to trigger exactly on usage of the APIs. Now it seems to trigger on a more complex logic. Probably some kind of heuristic to show it when an app does it the first time, but then only show it after a certain time if the app spam calls.

So in the case of AltTab, we see it only occasionally these days. This is what people here mostly wanted so i guess Apple improved things.

**Comment by aubreypwd** [comment]
If an app (like AltTab) needs to grab part of the screen (to generate a thumbnail for the selected window) it will report to macOS that it is recording your screen (because it is). 

This is normal behavior.

Doesn't mean I like the big giant purple square Apple chose (a dot would have sufficed), but I am still shocked this is still an issue. Disable thumbnails to make it go away.

**Comment by DownrightNifty** [comment]
> Disable thumbnails to make it go away.

I imagine for most people, myself included, the main appeal of the app is the thumbnails.

**Comment by DownrightNifty** [comment]
I found a way to allow user-specified apps to bypass the indicator without creating new security issues or disabling SIP, but unfortunately it's (tangentially) related to a security issue I reported to Apple, so I don't want to risk talking about it just yet. I'll post an update when I get a response from Apple. (The fix for this issue won't prevent the method I found from working.)

**Comment by mayeths** [comment]
Every month I check this thread for any new messages, and I still use MacOS 13.2 because of this. Disappointing to Apple for such a thing.

**Comment by DownrightNifty** [comment]
Hey all, Apple has unfortunately been quite slow to respond, so in the meantime I thought I would share another method of running AltTab without activating the indicator.

Paste this into Apple Terminal (not a third party terminal):
```
screen -dmS AltTab /Applications/AltTab.app/Contents/MacOS/AltTab
```
Then grant Terminal screen recording permission.

It's not automated, but it works for the time being. I'm still waiting on Apple and I'll provide a better solution when they get back to me.

**Comment by Duc-NSH** [comment]
> Hey all, Apple has unfortunately been quite slow to respond, so in the meantime I thought I would share another method of running AltTab without activating the indicator.
> 
> Paste this into Apple Terminal (not a third party terminal):
> 
> ```
> pkill AltTab; screen -dmS AltTab /Applications/AltTab.app/Contents/MacOS/AltTab
> ```
> 
> Then grant Terminal screen recording permission.
> 
> It's not automated, but it works for the time being. I'm still waiting on Apple and I'll provide a better solution when they get back to me.

So by manually giving the screen recording permission via the terminal, the system does not indicate that anymore?

**Comment by caff2code** [comment]
> I'll provide a better solution when they get back to me.

The one you provided already works perfectly, thank you! I wanted it to autostart with my mac so I made a `alttab.command` file and paste that in there (minus the `pkill AltTab`), then System preferences > Login items > + > `alttab.command`.

```bash
# alttab.command
screen -dmS AltTab /Applications/AltTab.app/Contents/MacOS/AltTab
```

`.command` is different to `.sh`, it will open in Apple's terminal always. Only minor caveat is you see the flash of a console window every reboot. But think of it as a reminder of the "good old days" of random background Windows command prompt flashes :)

As far as how it works, I guess Apple allows Apple's own apps to bypass the sharing icon, which is why, e.g. Screenshot.app doesn't activate it. Fingers crossed they don't remove it. I cannot see why they would do such a thing, because only advanced users would take advantage of this.

@lwouis Did you see this?

**Comment by conker-rsc** [comment]
Seconding the workaround - I added a startup trigger on BTT to execute the terminal command @Duc-NSH posted and it's been working perfectly. Thank you so much!

**Comment by arozbiz** [comment]
Does this disable the purple screen-recording notification icon or does it just keep it on permanently?

**Comment by conker-rsc** [comment]
> Does this disable the purple screen-recording notification icon or does it just keep it on permanently?

It prevents the purple icon from showing up at all

**Comment by jckpn** [comment]
For anyone else who was getting a permissions error, you might need to modify the file with `chmod 755 ~/alttab.command` (or whatever path to your command file).

<img width="305" alt="image" src="https://github.com/lwouis/alt-tab-macos/assets/14837124/f21ba3e8-f7d7-4b5c-8c58-a40cf4825da9">

**Comment by lockieluke** [comment]
I have a script for auto starting AltTab:

1. Create a file at `~/Library/LaunchAgents/local.alttab.plist` and paste this in:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
	<key>Label</key>
	<string>local.alttab</string>
	<key>ProgramArguments</key>
	<array>
		<string>/bin/sh</string>
		<string>-c</string>
		<string>/usr/bin/osascript $HOME/Scripts/alttab.scpt</string>
	</array>
	<key>RunAtLoad</key>
	<true/>
</dict>
</plist>
```
2. Make sure you turn off **Launch at login** in AltTab
3. Create a file at `~/Scripts/alttab.scpt` and paste this in:
```applescript
tell application "Terminal"
	do script "screen -dmS AltTab /Applications/AltTab.app/Contents/MacOS/AltTab"
    delay 1.5
    if it is running then quit
end tell
```
4. Run `launchctl load ~/Library/LaunchAgents/local.alttab.plist` to load the launch agent profile you just created
5. Logout and log back in, you might see the Terminal flash for a fraction of a second as it's starting AltTab but it should start working in a second

**Comment by mikluko** [comment]
> @lockieluke This solution doesn't seem to be working for me anymore. Is it still working for you, and if not, is there a new workaround?

Still working for me.

**Comment by mikluko** [comment]
Here is a bit improved version of @lockieluke's solution:

https://gist.github.com/mikluko/4618cc2f69bbb09d26d02f99a2897b66.js

How to use:

Boot out the existing launchd agent (if you have one):

```
launchctl bootout gui/$(id -u) ~/Library/LaunchAgents/local.alttab.plist
```

Install the new launchd agent:

```
curl -sl https://gist.githubusercontent.com/mikluko/4618cc2f69bbb09d26d02f99a2897b66/raw/b0e949de3c0171d936af4e932db489881944639e/local.alttab.plist > ~/Library/LaunchAgents/local.alttab.plist
launchctl bootstrap gui/$(id -u) ~/Library/LaunchAgents/local.alttab.plist
```

**Comment by nendonerd** [comment]
> > I'll provide a better solution when they get back to me.
> 
> The one you provided already works perfectly, thank you! I wanted it to autostart with my mac so I made a `alttab.command` file and paste that in there (minus the `pkill AltTab`), then System preferences > Login items > + > `alttab.command`.
> 
> ```shell
> # alttab.command
> screen -dmS AltTab /Applications/AltTab.app/Contents/MacOS/AltTab
> ```
> 
> `.command` is different to `.sh`, it will open in Apple's terminal always. Only minor caveat is you see the flash of a console window every reboot. But think of it as a reminder of the "good old days" of random background Windows command prompt flashes :)
> 
> As far as how it works, I guess Apple allows Apple's own apps to bypass the sharing icon, which is why, e.g. Screenshot.app doesn't activate it. Fingers crossed they don't remove it. I cannot see why they would do such a thing, because only advanced users would take advantage of this.
> 
> @lwouis Did you see this?

I'm using displaylink which also have this annoying flickering purple icon "xxx is recording" thing. I modify your code a bit, so that when the command finished running, the Terminal got killed. Also I have to specify the architecture else it will fail to execute. The app is a universal binary, I have no idea why I have to specify the architecture for it to run. 

```sh
# displaylink.command
# System preferences > Login items > + > `displaylink.command`
screen -dmS displaylink arch -arm64 /Applications/DisplayLink\ Manager.app/Contents/MacOS/DisplayLinkUserAgent; killall Terminal
```

**Comment by adamnovak** [comment]
@nendonerd's solution also works with the competing SiliconMotion InstantView application for their line of adapters, with the added caveat that you need to use the version of the app that lives in `Application Support`, which is the one that actually runs and does the capture. I tried to set up an AppleScript solution, but using Terminal.app seems to work better, because as soon as you export an AppleScript to something that can be a login item, it no longer gets to run as Script Editor.
```
# Add this as a login item
# Make sure Terminal is approved to do screen recording
screen -dmS instantview arch -arm64 "/Users/${USER}/Library/Application Support/InstantView/macOS InstantView.app/Contents/MacOS/macOS InstantView"
# Actually quit Terminal
# See <https://stackoverflow.com/a/22447960>
osascript -e "do shell script \"osascript -e \\\"tell application \\\\\\\"Terminal\\\\\\\" to quit\\\" &> /dev/null &\""
```

This is essential to prevent the screen recording icon from randomly flashing on and off as Apple's heuristics struggle to understand the concept of a userspace display adapter driver.

**Comment by jquick** [comment]
@nendonerd Does this work with the most recent 1.10+ version? Running manually via
```
/Applications/DisplayLink\ Manager.app/Contents/MacOS/DisplayLinkUserAgent
```
Errors with `Setting DisplayLink Login Item to false FAILED` and it starts kind of in a limited mode without permissions.

**Comment by Waitsnake** [comment]
Also had these days issues with the message "AltTab is sharing your screen" constantly popping up in a blue menubar item Sonoma 14.7 (23H124) even if I already grated AltTab the screen recording permissions in the past.

I don't know if i was just blind before or if it was the last Sonoma update I installed that started this? But I cannot remember seeing this blue menubar item popping up before like each minute or two? 😕

Anyway, I found a workaround here.

> 3. Create a file at `~/Scripts/alttab.scpt` and paste this in:
> 
> ```applescript
> tell application "Terminal"
> 	do script "screen -dmS AltTab /Applications/AltTab.app/Contents/MacOS/AltTab"
>     delay 1.5
>     if it is running then quit
> end tell
> ```

I just added this apple script under System preferences > Login items (it can handle applescripts directly) and fortunately it works.

**Comment by pulver22** [comment]
> Also had these days issues with the message "AltTab is sharing your screen" constantly popping up in a blue menubar item Sonoma 14.7 (23H124) even if I already grated AltTab the screen recording permissions in the past.
> 
> I don't know if i was just blind before or if it was the last Sonoma update I installed that started this? But I cannot remember seeing this blue menubar item popping up before like each minute or two? 😕
> 
> Anyway, I found a workaround here.
> 
> > 3. Create a file at `~/Scripts/alttab.scpt` and paste this in:
> > 
> > tell application "Terminal"
> > 	do script "screen -dmS AltTab /Applications/AltTab.app/Contents/MacOS/AltTab"
> >     delay 1.5
> >     if it is running then quit
> > end tell
> 
> I just added this apple script under System preferences > Login items (it can handle applescripts directly) and fortunately it works.

I've also tried this solution, but the 'Script Editor' app automatically opens every time I boot. Is it the same for you? The [solution](https://github.com/lwouis/alt-tab-macos/issues/2606#issuecomment-2042674340) from @lockieluke still seems the most silent one.

**Comment by evanrlong** [comment]
> > > I'll provide a better solution when they get back to me.
> 
> > 
> 
> > The one you provided already works perfectly, thank you! I wanted it to autostart with my mac so I made a `alttab.command` file and paste that in there (minus the `pkill AltTab`), then System preferences > Login items > + > `alttab.command`.
> 
> > 
> 
> > ```shell
> 
> > # alttab.command
> 
> > screen -dmS AltTab /Applications/AltTab.app/Contents/MacOS/AltTab
> 
> > ```
> 
> > 
> 
> > `.command` is different to `.sh`, it will open in Apple's terminal always. Only minor caveat is you see the flash of a console window every reboot. But think of it as a reminder of the "good old days" of random background Windows command prompt flashes :)
> 
> > 
> 
> > As far as how it works, I guess Apple allows Apple's own apps to bypass the sharing icon, which is why, e.g. Screenshot.app doesn't activate it. Fingers crossed they don't remove it. I cannot see why they would do such a thing, because only advanced users would take advantage of this.
> 
> > 
> 
> > @lwouis Did you see this?
> 
> 
> 
> I'm using displaylink which also have this annoying flickering purple icon "xxx is recording" thing. I modify your code a bit, so that when the command finished running, the Terminal got killed. Also I have to specify the architecture else it will fail to execute. The app is a universal binary, I have no idea why I have to specify the architecture for it to run. 
> 
> 
> 
> ```sh
> 
> # displaylink.command
> 
> # System preferences > Login items > + > `displaylink.command`
> 
> screen -dmS displaylink arch -arm64 /Applications/DisplayLink\ Manager.app/Contents/MacOS/DisplayLinkUserAgent; killall Terminal
> 
> ```
> 
> Also don't forget to make the file executable by `chmod +x displaylink.command`

Anyone know if this has been patched out by Apple?  I am able to run DisplayLink at Login using this method but still get the purple icons showing.

**Comment by leofrachet11** [comment]
@DownrightNifty, I know it’s been a while since your last update, but the workaround you shared already helped a lot of people, and I was just curious if the method you originally found still works, and if Apple ever got back to you on it?
